### ISSUE TITLE
Rule proposal: Prevent stray syntax from ending up in front of users

### ISSUE THREAD

**Issue by captbaritone** [jsx-eslint/eslint-plugin-react#3586]
JSX syntax makes it very easy to leave stray syntax behind while updating code. For example, say you have:

```javascript
return <h1>Hello!</h1>;
```

And then realize you want to wrap it in some other element. You might accidentally end up with:

```javascript
return (
  <Wrapper>
    <h1>Hello!</h1>;
  </Wrapper>
);
```

Which will leave a stray semicolon in your user-facing UI.

---

Does a rule already exist here, or elsewhere, to prevent that type of bug?

If not, I wrote one which looks like this:

```javascript
const SUSPECT_STRINGS = new Set([';', '}', ')}', ');']);

module.exports = {
  create(context) {
    return {
      'JSXText, Literal'(node) {
        if (
          (node.type !== 'JSXText' && node._babelType !== 'JSXText') ||
          typeof node.value !== 'string'
        ) {
          return;
        }
        if (!SUSPECT_STRINGS.has(node.value.trim())) {
          return;
        }
        context.report({node, message: 'Possible JSX Cruft'});
      },
    };
  },
};
```

Which uncovered hundreds of issues in our large code base with relatively few false positives.

**Comment by ljharb** [comment]
I'm glad you had success with your rule; I'm very skeptical a heuristics-based approach will actually provide few false positives in a general sense (and *any* false positives are largely unacceptable for a lint rule).

There's legit use cases for having a prose semicolon appear after a jsx element.

**Comment by ljharb** [comment]
What other examples besides a semicolon would you have in mind?

**Comment by captbaritone** [comment]
From the implementation I provided above, you can see that we're currently linting for `;`, `}`, `)}`, and `);`.

**Comment by captbaritone** [comment]
Anything I can do to help drive toward an outcome here? I believe this is could be a very impactful rule for the ecosystem.

**Comment by ljharb** [comment]
I’m willing to review test cases, if you’re willing to try some - docs and implementation are optional :-) i don’t want you to do unnecessary work if I’m not convinced, but it’s easiest to understand for me to review actual code.
